File: leetcode/lc-us/cpp/202.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    bool isHappy(int n) {
        unordered_set<int> occurred;
        while (n != 1) {
            int num = process(n);
            if (occurred.count(num)) {
                return false;
            }
            occurred.insert(num);
            n = num;
        }
        return true;
    }

private:
    int process(int num) {
        int sum = 0;
        while (num) {
            sum += (num % 10) * (num % 10);
            num /= 10;
        }
        return sum;
    }
};

File: leetcode/lc-us/rust/202.rs
use std::collections::HashSet;

impl Solution {
    pub fn is_happy(n: i32) -> bool {
        let mut set: HashSet<i32> = HashSet::new();
        let mut n = n;
        while n != 1 {
            let mut s = 0;
            while n > 0 {
                let d = n % 10;
                s += d * d;
                n /= 10;
            }
            if set.contains(&s) {
                return false;
            }
            set.insert(s);
            n = s;
        }
        true
    }
}